sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnarBatch.java

```diff
@@ -108,6 +111,170 @@ public InternalRow getRow(int rowId) {
 
   public ColumnarBatch(ColumnVector[] columns) {
     this.columns = columns;
-    this.row = new MutableColumnarRow(columns);
+    this.row = new ColumnarBatchRow(columns);
   }
 }
+
+/**
+ * An internal class, which wraps an array of {@link ColumnVector} and provides a row view.
+ */
+class ColumnarBatchRow extends InternalRow {
```

Comment:
Make it `final class` like `ColumnarRow` and `MutableColumnarRow`?

Reply:
yes we can add a final. I'll do it when I touch code here next time.